iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/granters/IamAuthorizationCodeTokenGranter.java

```diff
@@ -0,0 +1,106 @@
+import org.springframework.security.oauth2.provider.OAuth2Request;
+import org.springframework.security.oauth2.provider.OAuth2RequestFactory;
+import org.springframework.security.oauth2.provider.TokenRequest;
+import org.springframework.security.oauth2.provider.code.AuthorizationCodeServices;
+import org.springframework.security.oauth2.provider.token.AbstractTokenGranter;
+import org.springframework.security.oauth2.provider.token.AuthorizationServerTokenServices;
+
+import static it.infn.mw.iam.core.oauth.IamOAuth2RequestFactory.RESOURCE;
+
+// This is the org.springframework.security.oauth2.provider.code.AuthorizationCodeTokenGranter
+// class where the OAuth2Authentication request is created without the _resource_ parameter,
+// if it was not indicated in the token request
+@SuppressWarnings("deprecation")
+public class IamAuthorizationCodeTokenGranter extends AbstractTokenGranter {
+
+  private static final String GRANT_TYPE = "authorization_code";
+
+  private final AuthorizationCodeServices authorizationCodeServices;
+
+  public IamAuthorizationCodeTokenGranter(AuthorizationServerTokenServices tokenServices,
+      AuthorizationCodeServices authorizationCodeServices,
+      ClientDetailsService clientDetailsService, OAuth2RequestFactory requestFactory) {
+    super(tokenServices, clientDetailsService, requestFactory, GRANT_TYPE);
+    this.authorizationCodeServices = authorizationCodeServices;
+  }
+
+  @Override
+  protected OAuth2Authentication getOAuth2Authentication(ClientDetails client,
+      TokenRequest tokenRequest) {
+
+    Map<String, String> parameters = tokenRequest.getRequestParameters();
+    String authorizationCode = parameters.get("code");
+    String redirectUri = parameters.get(OAuth2Utils.REDIRECT_URI);
+
+    if (authorizationCode == null) {
+      throw new InvalidRequestException("An authorization code must be supplied.");
+    }
+
+    OAuth2Authentication storedAuth =
+        authorizationCodeServices.consumeAuthorizationCode(authorizationCode);
+    if (storedAuth == null) {
+      throw new InvalidGrantException("Invalid authorization code");
+    }
+
+    OAuth2Request pendingOAuth2Request = storedAuth.getOAuth2Request();
+    String redirectUriApprovalParameter =
+        pendingOAuth2Request.getRequestParameters().get(OAuth2Utils.REDIRECT_URI);
+
+    if ((redirectUri != null || redirectUriApprovalParameter != null)
+        && !pendingOAuth2Request.getRedirectUri().equals(redirectUri)) {
+      throw new RedirectMismatchException("Redirect URI mismatch.");
+    }
+
+    String pendingClientId = pendingOAuth2Request.getClientId();
+    String clientId = tokenRequest.getClientId();
+    if (clientId != null && !clientId.equals(pendingClientId)) {
+      throw new InvalidClientException("Client ID mismatch");
+    }
+
+    Map<String, String> combinedParameters =
+        new HashMap<String, String>(pendingOAuth2Request.getRequestParameters());
+    combinedParameters.putAll(parameters);
```

Comment:
Not sure why we have to include the authorization request parameters instead of the token request ones at the token creation, but it was the default behavior in Spring.